server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs

```diff
@@ -0,0 +1,153 @@
+use repository::{
+    EqualFilter, InvoiceLineFilter, InvoiceLineRepository, InvoiceLineType, RepositoryError,
+    StorageConnection,
+};
+use util::uuid;
+
+use crate::invoice_line::{
+    outbound_shipment_unallocated_line::{
+        DeleteOutboundShipmentUnallocatedLine, InsertOutboundShipmentUnallocatedLine,
+        UpdateOutboundShipmentUnallocatedLine,
+    },
+    stock_out_line::{DeleteStockOutLine, InsertStockOutLine, StockOutType, UpdateStockOutLine},
+};
+
+use super::{SaveOutboundShipmentItemLines, SaveOutboundShipmentLine};
+
+pub enum ManagePlaceholderLine {
+    Insert(InsertOutboundShipmentUnallocatedLine),
+    Update(UpdateOutboundShipmentUnallocatedLine),
+    Delete(DeleteOutboundShipmentUnallocatedLine),
+    NothingToDo,
+}
+pub struct GenerateResult {
+    pub lines_to_add: Vec<InsertStockOutLine>,
+    pub lines_to_update: Vec<UpdateStockOutLine>,
+    pub lines_to_delete: Vec<DeleteStockOutLine>,
+    pub manage_placeholder: ManagePlaceholderLine,
+}
+
+pub fn generate(
+    connection: &StorageConnection,
+    SaveOutboundShipmentItemLines {
+        invoice_id,
+        item_id,
+        lines,
+        placeholder_quantity,
+    }: SaveOutboundShipmentItemLines,
+) -> Result<GenerateResult, RepositoryError> {
+    let existing_lines = InvoiceLineRepository::new(connection).query_by_filter(
+        InvoiceLineFilter::new()
+            .item_id(EqualFilter::equal_to(&item_id))
+            .invoice_id(EqualFilter::equal_to(&invoice_id)),
+    )?;
+
+    let check_already_exists = |id: &str| {
+        existing_lines
+            .iter()
+            .any(|line| line.invoice_line_row.id == id)
+    };
+
+    let lines_to_add = lines
+        .clone()
+        .into_iter()
+        .filter(|line| line.number_of_packs > 0.0 && !check_already_exists(&line.id))
```

Comment:
hmm, so for the 'ids'for draft lines, I think we should really go with stock_line_ids (when getting draft lines and here when checking if there is invoice line)